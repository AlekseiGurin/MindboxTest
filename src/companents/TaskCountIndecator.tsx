import React from 'react';
import { DisplayedTasksType, TodoType } from "./Todo";
import { Spin } from "antd"; 

interface TaskCountIndicatorProps {
    displayedTasks: DisplayedTasksType,
    todos?: TodoType[]
}

export const TaskCountIndicator: React.FC<TaskCountIndicatorProps> = ({ displayedTasks, todos }) => {
    const getContent = () => {
        let content;
        switch (displayedTasks) {
            case "All":
            case "Active":
                content = `${todos?.filter(item => !item.done).length} items left`;
                break;
            case "Complited": 
                content = `${todos?.filter(item => item.done).length} already done`;
                break;
        }
        return <span data-testid="countIndicator">{content}</span>
    } 
    return (
        <div className="task-count-indicator">
            { todos ? getContent() : <Spin/> }
        </div>
    )
}