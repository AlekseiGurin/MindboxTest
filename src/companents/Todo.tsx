import React, { useEffect, useMemo, useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import { Button, Input, Form, Skeleton } from "antd"
;
import { TodoList } from './TodoList';
import { DownOutlined, UpOutlined } from '@ant-design/icons';
import { TaskCountIndicator } from './TaskCountIndecator';

let tasks = [
    {
        tadoText: "Тестовое задание",
        done: false,
        id: uuidv4(),
    },
    {
        tadoText: "Прекрасный код",
        done: true,
        id: uuidv4(),
    },
    {
        tadoText: "Покрытие тестами",
        done: false,
        id: uuidv4(),
    }
];

export type TodoType = {
    tadoText: string,
    done: boolean,
    id: string
}

export type DisplayedTasksType = "All" | "Active" | "Complited";

export const Todo = () => {

const [form] = Form.useForm();
const [defaultTodos, setDefaultTodos] = useState<TodoType[]>();
const [displayedTasks, setDisplayedTasks] = useState<DisplayedTasksType>("All");
const [visibleTodolist, setVisibleTodoList] = useState(true);

useEffect(() => {
   setTimeout(() => {
    setDefaultTodos(tasks || []);
   }, 1000);
}, []);

const todos = useMemo(() => {
    switch (displayedTasks) {
        case "All":
            return defaultTodos;
        case "Active":
            return defaultTodos?.filter(item => !item.done);
        case "Complited": 
            return defaultTodos?.filter(item => item.done);
    };
}, [displayedTasks, defaultTodos]);

const onButtonTypeTaskKlick = (activeButton: DisplayedTasksType) => {
    setDisplayedTasks(activeButton);
    
};

const onClear = () => {
    const clearedList = defaultTodos?.filter(item => !item.done);
    console.log("clearedList", clearedList);
    setDefaultTodos(clearedList);
};

const onCompletedTodo = (todoId: string) => {
    const newList = defaultTodos?.map(item => {
        const todo = { ...item };
        if(todo.id === todoId) {
          todo.done = !todo.done
        }
        return todo;
    });
    setDefaultTodos(newList);
};

const onChangeTodos = (e: { newTodo: string }) => {
    console.log("onChangeTodos e", e);
    const newTodo = {
        tadoText: e.newTodo,
        done: false,
        id: uuidv4(),
    }
    todos && setDefaultTodos([...todos, newTodo]);
    form.resetFields();
};

    return (
        <div className="todo-desk">
            <div className="label">todos</div>
            <div className="content-container">
                <div className="input-container">
                    <Button data-testid="hiddTodoListBtn" onClick={() => setVisibleTodoList(!visibleTodolist)} type="text">{visibleTodolist ? <DownOutlined /> : <UpOutlined />}</Button>
                    <Form
                        form={form}
                        onFinish={onChangeTodos}
                    >
                        <Form.Item name="newTodo" rules={[{ required: true, message: "todo must not be an empty string" }]}>
                            <Input disabled={!todos} placeholder="What needs to be done?" />
                        </Form.Item>
                    </Form>
                </div>
                { todos ? <TodoList todos={todos} onCompletedTodo={onCompletedTodo} visibleTodolist={visibleTodolist} /> : <Skeleton />}  
                <div className="todo-desk-footer">
                    <TaskCountIndicator displayedTasks={displayedTasks} todos={todos} />
                    <div className="filter-button-block">
                        <Button disabled={!todos} key="All" onClick={() => onButtonTypeTaskKlick("All")} type={displayedTasks === "All" ? "default" : "text"}>All</Button>
                        <Button disabled={!todos} key="Active" onClick={() => onButtonTypeTaskKlick("Active")} type={displayedTasks === "Active" ? "default" : "text"}>Active</Button>
                        <Button data-testid="complitedBtn" disabled={!todos} key="Complited" onClick={() => onButtonTypeTaskKlick("Complited")} type={displayedTasks === "Complited" ? "default" : "text"}>Complited</Button>
                    </div>
                    <Button disabled={!todos} className="clear-button"onClick={() => onClear()} type="text">Clear complited</Button>
                </div>
            </div>
        </div>
    )
}