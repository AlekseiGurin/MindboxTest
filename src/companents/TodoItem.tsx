import React from 'react';
import { TodoType } from './Todo';
import { CheckOutlined } from '@ant-design/icons';

export const TodoItem: React.FC<TodoType> = ({ done, tadoText, id }) => {
    return (
        <div key={id} id={id} className="todo-item">
            <div className="circle">{done && <CheckOutlined />}</div>
            <div className={done ? "tado-text done" : "tado-text"}>{tadoText}</div>
        </div>
    )
}