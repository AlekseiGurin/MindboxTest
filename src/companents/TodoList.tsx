import React from 'react';
import { TodoType } from './Todo';
import { CheckOutlined } from '@ant-design/icons';
import { Empty, Button } from 'antd';

export interface TodoListProps {
    todos: TodoType[];
    onCompletedTodo: (todoId: string) => void;
}

export const TodoList: React.FC<TodoListProps> = ({ todos, onCompletedTodo }) => {
    return (
        <div data-testid="todoList" className="todos">
           {todos.length ? todos.map(item => (
                <div onClick={() => onCompletedTodo(item.id)} id={item.id} key={item.id} className="todo-item">
                    <div className="circle">{item.done && <CheckOutlined />}</div>
                    <div className={item.done ? "tado-text done" : "tado-text"}>{item.tadoText}</div>
                </div>
            )) : <Empty description="empty list" />}
        </div>
    )
}