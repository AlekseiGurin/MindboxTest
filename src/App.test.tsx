import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Todo } from "./companents/Todo";
import { TodoList } from "./companents/TodoList";
import { TaskCountIndicator } from "./companents/TaskCountIndecator";


describe("Todo items render", () => {
    test("desk label render", () => {
        render(<Todo />);
        const deskLabel = screen.getByText(/todos/i);
        expect(deskLabel).toBeInTheDocument();
      });

    test("input event", () => {
        render(<Todo />);
        const input = screen.getByPlaceholderText(/What needs to be done/i);
        expect(input).toMatchSnapshot();
        fireEvent.input(input, {
            EventTarget: {
                value: "new todo!"
            }
        });
        const newTodo = screen.queryAllByText(/new todo!/i);
        expect(newTodo).toBeInTheDocument;
    });

    test("task count indicator render", async () => {
        render(<TaskCountIndicator todos={[]} displayedTasks="All" />);
        const taskCountIndicator = await screen.findByTestId("countIndicator");
        expect(taskCountIndicator).toBeInTheDocument();
    });

    test("todo list render", async () => {
        const fnc = jest.fn();
        render(<TodoList todos={[]} onCompletedTodo={fnc} />);
        const todoList = await screen.findByTestId("todoList");
        expect(todoList).toBeInTheDocument;
    });
    
    test("onComplited render", () => {
        render(<Todo />);
        const btn = screen.getByTestId("complitedBtn");
        const complitedCounter = screen.queryAllByText(/already done/i);
        expect(complitedCounter).toBeNull;
        fireEvent.click(btn);
        expect(complitedCounter).toBeInTheDocument;
    });
});