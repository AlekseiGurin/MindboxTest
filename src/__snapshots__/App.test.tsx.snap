// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo items render desk input render 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`Todo items render input event 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`renders App items 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`renders App items render  1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`renders App items render desk input 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`renders App items render desk label 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`renders App items renders App items 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`renders Todo items  desk input 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`renders Todo items render desk input 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`renders desk label 1`] = `
<input
  aria-required="true"
  class="ant-input ant-input-disabled css-dev-only-do-not-override-1htoz2s ant-input-outlined"
  disabled=""
  id="newTodo"
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;
